src/WebSocketManager/WebSocketManagerMiddleware.cs

```diff
@@ -31,27 +27,39 @@ public class WebSocketManagerMiddleware
                 return;
             
             var socket = await context.WebSockets.AcceptWebSocketAsync();
-            _webSocketManager.AddSocket(socket);
-
-            var socketId = _webSocketManager.GetId(socket);
-            await _webSocketMessageHandler.SendMessageToAllAsync(message: $"{_webSocketManager.GetId(socket)} connected");
-            
+            await _webSocketHandler.OnConnected(socket);
             
-            await _webSocketMessageHandler.ReceiveMessageAsync(socket, async (result, buffer) => {
-
+            await Receive(socket, async(result, buffer) =>
+            {
                 if(result.MessageType == WebSocketMessageType.Text)
                 {
-                    var message = $"{socketId} said: {Encoding.UTF8.GetString(buffer, 0, result.Count)}";
-                    await _webSocketMessageHandler.SendMessageToAllAsync(message: message);
+                    await _webSocketHandler.ReceiveAsync(socket, result, buffer);
+                    return;
                 }
 
-                if(result.MessageType == WebSocketMessageType.Close)
+                else if(result.MessageType == WebSocketMessageType.Close)
                 {
-                    await socket.CloseOutputAsync(WebSocketCloseStatus.NormalClosure, String.Empty, CancellationToken.None);
+                    await _webSocketHandler.OnDisconnected(socket);
+                    return;
                 }
+
             });
             
+            //TODO - investigate the Kestrel exception thrown when this is the last middleware
```

Comment:
Investigate Kestrel exception - is it only thrown when middleware is last on chain?